File: app/src/lib/klingelton.ts
/**
 * Der Klingelton: welcher Klang auf welchen Boxen spielt, wenn es klingelt.
 *
 * Gehört zur Regel «Es klingelt» (core/notifyrules.py) und steht deshalb in
 * deren Karte, wie die Storen-Auswahl der Wächter-Regeln in ihrer eigenen.
 * Anders als dort heisst eine leere Auswahl hier «still», nicht «alle
 * Boxen» - ein Klingelton soll erst losgehen, wenn ihn jemand eingerichtet
 * hat, nicht beim ersten Klingeln nach der Auslieferung auf jeder Box im
 * Haus (core/klingelton.py).
 */

export interface Lautsprecher {
  id: string;
  name: string;
  room?: string | null;
}

export interface Klang {
  key: string;
  label: string;
}

/**
 * Eine gewählte Box - mit ihrer eigenen Lautstärke und Zeitspanne.
 *
 * Beides gehört je Box und nicht ins Haus: Die Küchenbox steht neben dem
 * Esstisch und darf leise sein, im Keller hört man sonst nichts; und die
 * Box im Kinderzimmer soll abends nicht mehr losgehen, während die im
 * Flur immer darf. Der Hub liest beide Formen - eine blosse Kennung aus
 * einer älteren Fassung bekommt die Vorgaben (core/klingelton.py).
 */
export interface Klingelbox {
  id: string;
  volume: number;
  /** «07:00» - von wann an es auf dieser Box klingelt. */
  from: string;
  /** «20:00», oder «24:00» für «bis Mitternacht». */
  to: string;
}

export interface Klingeltonstand {
  sound: string;
  speakers: Klingelbox[];
  sounds: Klang[];
  candidates: Lautsprecher[];
}

/** Lautstärke einer Box, solange niemand daran gedreht hat. */
export const LAUTSTAERKE_VORGABE = 55;
/** Und ihre Zeitspanne: den ganzen Tag. Wer eine Box wählt, will sie hören. */
export const SPANNE_VORGABE = { from: '00:00', to: '24:00' };

/** Ist diese Box gewählt? (rein, testbar) */
export function istGewaehlt(speakers: Klingelbox[], id: string): boolean {
  return speakers.some((box) => box.id === id);
}

/** Die Einstellungen einer Box - oder die Vorgaben (rein, testbar). */
export function boxStand(speakers: Klingelbox[], id: string): Klingelbox {
  return (
    speakers.find((box) => box.id === id) ?? {
      id,
      volume: LAUTSTAERKE_VORGABE,
      ...SPANNE_VORGABE,
    }
  );
}

/** Name und Raum in einer Zeile (rein, testbar). */
export function lautsprecherName(box: Lautsprecher): string {
  const raum = (box.room ?? '').trim();
  if (!raum || box.name.toLowerCase().includes(raum.toLowerCase())) {
    return box.name;
  }
  return `${box.name} · ${raum}`;
}

/**
 * Eine Box aus der Auswahl nehmen oder dazunehmen (rein, testbar).
 *
 * Eine neue Box kommt mit den Vorgaben herein. Eine abgewählte verliert
 * ihre Einstellungen - das ist gewollt: Wer sie später wieder dazunimmt,
 * fängt sichtbar bei den Vorgaben an, statt eine halb vergessene
 * Nachtsperre von vor drei Monaten zu erben.
 */
export function boxUmschalten(gewaehlt: Klingelbox[], id: string): Klingelbox[] {
  return istGewaehlt(gewaehlt, id)
    ? gewaehlt.filter((box) => box.id !== id)
    : [...gewaehlt, { id, volume: LAUTSTAERKE_VORGABE, ...SPANNE_VORGABE }];
}

/** Eine Einstellung einer einzelnen Box ändern (rein, testbar). */
export function boxAendern(
  gewaehlt: Klingelbox[],
  id: string,
  aenderung: Partial<Omit<Klingelbox, 'id'>>
): Klingelbox[] {
  return gewaehlt.map((box) => (box.id === id ? { ...box, ...aenderung } : box));
}

/** Die Lautstärken zur Wahl - dieselbe Leiter wie beim Wecker. */
export const LAUTSTAERKEN = [20, 40, 55, 70, 85];

/**
 * Getipptes zu einer Uhrzeit machen (rein, testbar).
 *
 * «7» wird «07:00», «730» wird «07:30», «7:5» wird «07:05». Was gar
 * nicht geht, fällt auf die Vorgabe zurück statt die Box stumm zu
 * schalten: Eine kaputte Eingabe darf dazu führen, dass es zu oft
 * klingelt, nicht dass es nie klingelt (dieselbe Regel wie im Hub,
 * core/klingelton.py).
 */
export function uhrzeitSauber(text: string, vorgabe: string): string {
  const roh = (text ?? '').trim();
  if (roh === '24:00' || roh === '2400') return '24:00';
  const ziffern = roh.replace(/[^0-9]/g, '');
  // Ohne eine einzige Ziffer ist es keine Uhrzeit, sondern ein Wort.
  // Ohne diese Zeile wurde aus «abends» ein «00:00» - und damit aus
  // einem Vertipper still Mitternacht.
  if (!ziffern) return vorgabe;
  let stunde: number;
  let minute: number;
  if (roh.includes(':')) {
    const [links, rechts] = roh.split(':');
    stunde = Number(links);
    minute = Number(rechts === '' ? 0 : rechts);
  } else if (ziffern.length <= 2) {
    stunde = Number(ziffern);
    minute = 0;
  } else {
    stunde = Number(ziffern.slice(0, ziffern.length - 2));
    minute = Number(ziffern.slice(-2));
  }
  if (!Number.isInteger(stunde) || !Number.isInteger(minute)) return vorgabe;
  if (stunde < 0 || stunde > 23 || minute < 0 || minute > 59) return vorgabe;
  return `${String(stunde).padStart(2, '0')}:${String(minute).padStart(2, '0')}`;
}

/**
 * Was die Zeitspanne einer Box bedeutet, in einem Satz (rein, testbar).
 *
 * «00:00 bis 24:00» als Zahlenpaar dastehen zu lassen wäre die
 * schlechtere Auskunft: Wer es liest, rechnet nach, ob das nun immer
 * heisst oder nie.
 */
export function spanneSatz(box: Klingelbox): string {
  if (box.from === box.to || (box.from === '00:00' && box.to === '24:00')) {
    return 'immer';
  }
  return `${box.from} – ${box.to === '24:00' ? '24:00' : box.to}`;
}

/**
 * Was die Wahl gerade bewirkt, in einem Satz (rein, testbar).
 *
 * Ohne Box steht hier bewusst, was *nicht* passiert - sonst sucht man
 * den versprochenen Ton und hält sein Ausbleiben für einen Fehler.
 */
export function klingeltonSatz(stand: Klingeltonstand | null): string {
  if (!stand || stand.speakers.length === 0) {
    return 'Keine Box gewählt – es bleibt bei der Push-Nachricht allein.';
  }
  const klang = stand.sounds.find((eintrag) => eintrag.key === stand.sound)?.label ?? stand.sound;
  const namen = stand.speakers
    .map((eintrag) => stand.candidates.find((box) => box.id === eintrag.id))
    .filter((box): box is Lautsprecher => box != null)
    .map(lautsprecherName);
  const boxenText =
    namen.length <= 2 ? namen.join(' und ') : `${namen.length} Boxen`;
  return `«${klang}» spielt auf ${boxenText}.`;
}

File: hub/homepilot/core/klingelton.py
"""Der Ton, der auf gewählten Lautsprechern spielt, wenn es klingelt.

Der Tonmeister (core/ton.py) dämpfte beim Klingeln bisher nur eine
laufende Wiedergabe - wer gerade keine Musik hörte, bekam nichts als
die stille Push-Nachricht. Hier kommt ein eigener, hörbarer Ton dazu.

Eingebaut statt eingekauft: ein paar Sekunden Sinuston aus Frequenz und
Dauer brauchen keine Lizenz, kein Internet und keine Datei im Abbild -
sie entstehen beim Klingeln aus reinen Zahlen. Das erlaubt auch, ein
paar zum Schmunzeln dazuzunehmen: «wer klingelt» darf auch mal die
Quietscheente sein, nicht nur der klassische Gong.

Gespielt wird über denselben Weg wie eine Durchsage (core/say.py:
play_audio) - kurzlebige Adresse, play_url auf die Box, danach der alte
Zustand zurück. Der eigentliche Anstoss (welcher Ton, auf welchen
Boxen, überhaupt) steht in ``Tonmeister.klingelton_abspielen``.

Bewusst *ohne* Vorgabe-Lautsprecher: Ohne gewählte Box bleibt es still.
Bei den Storen-Wächtern (core/storenwaechter.py) heisst eine leere Wahl
«alle», weil dort ein Unwetter alle schützen soll. Ein Klingelton ist
kein Sicherheitsfall - er soll erst losgehen, wenn jemand ihn wirklich
eingerichtet hat, nicht beim ersten Klingeln nach der Auslieferung auf
jeder Box im Haus.
"""

from __future__ import annotations

import math
import struct
from typing import Any

#: Eine Note: (Frequenz in Hz, Dauer in Sekunden). 0 Hz heisst Stille -
#: eine Pause zwischen zwei Tönen, kein Ton bei 0 Hz.
Note = tuple[float, float]

SAMPLERATE = 22050

#: Lautstärke des Klingeltons in Prozent - hörbar, aber kein Schreck.
#: Etwas leiser als eine Durchsage (DURCHSAGE_VOLUME in core/say.py):
#: Eine Durchsage hat einen Satz zu sagen, der Klingelton nur «da ist
#: wer».
LAUTSTAERKE = 55

#: Wo die Wahl (Ton, Boxen) in der Datendatei liegt.
DATA_KEY = "klingelton"

#: Die eingebauten Klänge, in Anzeige-Reihenfolge. Die ersten drei sind
#: klassisch, die letzten drei zum Schmunzeln - beides darf man wollen.
KLAENGE: list[dict[str, Any]] = [
    {
        "key": "dingdong",
        "label": "Ding-Dong",
        "noten": [(784.0, 0.35), (0.0, 0.05), (523.0, 0.55)],
    },
    {
        "key": "dreiklang",
        "label": "Dreiklang",
        "noten": [(523.0, 0.18), (659.0, 0.18), (784.0, 0.42)],
    },
    {
        "key": "kuckuck",
        "label": "Kuckuck",
        "noten": [
            (659.0, 0.3),
            (0.0, 0.08),
            (523.0, 0.3),
            (0.0, 0.2),
            (659.0, 0.3),
            (0.0, 0.08),
            (523.0, 0.3),
        ],
    },
    {
        "key": "hupe",
        "label": "Hupe",
        "noten": [(220.0, 0.22), (0.0, 0.07), (220.0, 0.3)],
    },
    {
        "key": "quietscheente",
        "label": "Quietscheente",
        "noten": [
            (1150.0, 0.07),
            (850.0, 0.06),
            (1250.0, 0.07),
            (900.0, 0.06),
            (1400.0, 0.12),
        ],
    },
    {
        "key": "tusch",
        "label": "Tusch",
        "noten": [(523.0, 0.12), (659.0, 0.12), (784.0, 0.12), (1047.0, 0.4)],
    },
]

BY_KEY: dict[str, dict[str, Any]] = {klang["key"]: klang for klang in KLAENGE}

#: Der Ton, der gilt, solange niemand einen anderen gewählt hat.
STANDARD = "dingdong"

#: Die Note des Eingangs-Countdowns (Punkt 487 der Werkbank).
#:
#: Ein einzelner kurzer Piep, hoch genug, um durch eine geschlossene Tür
#: zu kommen, und kurz genug, dass er in einen Sekundentakt passt.
PIEP_HZ = 1046.0
PIEP_DAUER = 0.12

#: Wie eng die Piepser gegen Ende zusammenrücken.
#:
#: Der Ton soll nicht nur sagen «die Uhr läuft», sondern «sie läuft
#: ab» - und das ist der Unterschied zwischen einem Wecker und einer
#: Anlage. Zwei Sekunden Abstand am Anfang, eine halbe am Schluss: Wer
#: zur Tür hereinkommt, hört an der Dichte, wie viel Zeit bleibt, ohne
#: etwas anzusehen.
PIEP_ABSTAND_START = 2.0
PIEP_ABSTAND_ENDE = 0.5


def countdown_noten(sekunden: float) -> list[Note]:
    """Der Ton der Eingangsverzögerung (rein, testbar) - Punkt 487.

    Die Verzögerung lief im Hub korrekt ab und war nur zu sehen, wer die
    App öffnete - dann sind zehn der dreissig Sekunden weg. Ein Ton, der
    schneller wird, sagt dasselbe ohne Bildschirm und ohne Hände.

    Ein Stück Stille am Schluss bleibt bewusst weg: Die letzte Sekunde
    gehört dem Alarm, nicht dem Countdown.
    """
    dauer = max(0.0, float(sekunden))
    noten: list[Note] = []
    vergangen = 0.0
    while True:
        anteil = min(1.0, vergangen / dauer) if dauer > 0 else 1.0
        abstand = PIEP_ABSTAND_START + anteil * (PIEP_ABSTAND_ENDE - PIEP_ABSTAND_START)
        if vergangen + PIEP_DAUER + abstand > dauer:
            break
        noten.append((PIEP_HZ, PIEP_DAUER))
        noten.append((0.0, abstand))
        vergangen += PIEP_DAUER + abstand
    # Die letzte Pause weg: Sie würde die Wiedergabe künstlich in die
    # Sekunde hineinziehen, in der der Alarm losgeht.
    if noten and noten[-1][0] == 0.0:
        noten.pop()
    return noten


def ton_samples(frequenz: float, dauer: float, samplerate: int = SAMPLERATE) -> list[float]:
    """Ein Sinuston als Werte zwischen -1 und 1 (rein, testbar).

    Mit kurzer Ein- und Ausblendung (5 ms): Ohne sie knackt jeder Ton
    beim Einsetzen und Enden, hörbar als Klicken - besonders bei der
    Quietscheente, die aus lauter kurzen Tönen besteht.
    """
    anzahl = max(1, int(round(dauer * samplerate)))
    if frequenz <= 0:
        return [0.0] * anzahl
    fade = max(1, min(anzahl // 4, int(0.005 * samplerate)))
    werte: list[float] = []
    for i in range(anzahl):
        huellkurve = 1.0
        if i < fade:
            huellkurve = i / fade
        elif i >= anzahl - fade:
            huellkurve = (anzahl - i) / fade
        werte.append(huellkurve * math.sin(2 * math.pi * frequenz * i / samplerate))
    return werte


def noten_zu_samples(noten: list[Note], samplerate: int = SAMPLERATE) -> list[float]:
    """Eine Notenfolge aneinandergehängt (rein, testbar)."""
    samples: list[float] = []
    for frequenz, dauer in noten:
        samples.extend(ton_samples(frequenz, dauer, samplerate))
    return samples


def wav_bytes(samples: list[float], samplerate: int = SAMPLERATE) -> bytes:
    """16-Bit-Mono-WAV aus Werten zwischen -1 und 1 (rein, testbar).

    Von Hand statt mit dem ``wave``-Modul zusammengesetzt: Das Format
    ist für einen kurzen Mono-Ton einfach genug, und so bleibt die
    ganze Erzeugung ohne Seiteneffekt - kein Öffnen einer Datei, kein
    Zwischenspeicher, nur Bytes aus Zahlen.
    """
    frames = b"".join(
        struct.pack("<h", max(-32767, min(32767, int(round(wert * 32767)))))
        for wert in samples
    )
    kanaele, bits = 1, 16
    block_align = kanaele * bits // 8
    byte_rate = samplerate * block_align
    header = struct.pack(
        "<4sI4s4sIHHIIHH4sI",
        b"RIFF",
        36 + len(frames),
        b"WAVE",
        b"fmt ",
        16,
        1,
        kanaele,
        samplerate,
        byte_rate,
        block_align,
        bits,
        b"data",
        len(frames),
    )
    return header + frames


def klang_wav(key: str) -> bytes:
    """Der fertige Klingelton zu einem Schlüssel - die Vorgabe bei einem
    unbekannten (rein genug: rechnet nur, greift auf nichts zu)."""
    klang = BY_KEY.get(key) or BY_KEY[STANDARD]
    return wav_bytes(noten_zu_samples(klang["noten"]))


def einstellung_lesen(rows: Any) -> dict[str, Any]:
    """Die gespeicherte Wahl - Ton und Boxen (rein, testbar).

    ``rows`` ist der Datenspeicher-Eintrag (höchstens ein Dict in einer
    Liste, wie beim Gute-Nacht-Knopf). Ein unbekannter Ton fällt auf
    die Vorgabe zurück, damit eine spätere Auslieferung, die einen
    Klang umbenennt, niemanden mit einer kaputten Einstellung
    zurücklässt. Keine Boxen gewählt heisst still - siehe Kopf dieser
    Datei.
    """
    for row in rows or []:
        if isinstance(row, dict):
            sound = str(row.get("sound") or STANDARD)
            if sound not in BY_KEY:
                sound = STANDARD
            boxen = row.get("speakers")
            return {
                "sound": sound,
                "speakers": (
                    [str(eintrag) for eintrag in boxen if str(eintrag)]
                    if isinstance(boxen, list)
                    else []
                ),
            }
    return {"sound": STANDARD, "speakers": []}
